### DISCUSSION TITLE
Use freezed on class that has a Color member in it

### DISCUSSION THREAD

**Discussion by ToddManUtah** [rrousselGit/freezed#945] [Q&A] [answered]
I want to create a model called "Categories" and make it compatible with to/fromJson. It will have 2 fields: String name and Color color. When I use freezed, I can get the following code to generate:

    import 'dart:ui';
    import 'package:flutter/material.dart';
    import 'package:freezed_annotation/freezed_annotation.dart';
    
    part 'category.freezed.dart';
    part 'category.g.dart';
    
    @freezed
    class Category with _$Category {
      const factory Category({
        required String name,
        @JsonKey(includeFromJson: false, includeToJson: false) Color? color,
      }) = _Category;
    
      factory Category.fromJson(Map<String, dynamic> json) => _$CategoryFromJson(json);
    }

I have 2 problems:
1) I'd like to assign a default color to "color" and not make it optional. I can't seem to get this to work.
2) I'd also like to remove the named parameters. But when I remove the curly brackets, I get the following error:

code: 

    import 'dart:ui';
    import 'package:flutter/material.dart';
    import 'package:freezed_annotation/freezed_annotation.dart';
    
    part 'category.freezed.dart';
    part 'category.g.dart';
    
    @freezed
    class Category with _$Category {
      const factory Category(
        required String name,
        @JsonKey(includeFromJson: false, includeToJson: false) Color? color,
      ) = _Category;
    
      factory Category.fromJson(Map<String, dynamic> json) => _$CategoryFromJson(json);
    }

error: 
This builder requires Dart inputs without syntax errors.
However, package:shopping_list/models/category.dart (or an existing part) contains the following errors.
category.dart:24:5: Can't have modifier 'required' here.

If I remove the "required" in front of String name, I get the following error:
Cannot populate the required constructor argument: color. It is assigned to a field not meant to be used in fromJson.
package:shopping_list/models/category.freezed.dart:114:9
    ╷
114 │   const _$_Category(this.name,

Like I said, I'd like to exclude Color color from the to/fromJson but I'd like to give it a default value so that it's not optional. Any ideas?

**Comment by rrousselGit** [comment]
The problem is that your `name` parameter is "required" but it isn't "named". You're missing the `{}`

Do:

```dart
const factory Category({required String name, ...}) = _Category;`

**Reply by ToddManUtah** [reply]
Is there a way of doing this WITHOUT using named parameters? In other words, I'd like to use positional arguments/parameters:

ie: Category('meat') & Category('dairy', Colors.blue) VS  Category(name: 'meat') & Category(name: 'dairy', color: Colors.blue)

When I remove the curly braces AND remove 'required' from in front of name, I get the following error:
Cannot populate the required constructor argument: color. It is assigned to a field not meant to be used in fromJson.
package:shopping_list/models/category.freezed.dart:114:9
    ╷
114 │   const _$_Category(this.name,

**Reply by rrousselGit** [reply, marked as answer]
Don't specify the `required` keyword then. The error you shared initially has nothing to do with `Color` or Freezed. It's simply not valid Dart code

As for `fromJson`, Color is not serializable. If you want a from/toJson, you'll need a custom JsonConverter

**Reply by ToddManUtah** [reply]
BTW: Thanks for all the work you do. I've been using Riverpod extensively and have trained my whole development team how and where to use it. I thought I'd experiment with Freezed and it's really awesome as well. Thanks again for all your hard work.
